src/Features/Core/Portable/ConvertToInterpolatedString/AbstractConvertConcatenationToInterpolatedStringRefactoringProvider.cs

```diff
@@ -112,22 +111,45 @@ private Task<Document> UpdateDocumentAsync(Document document, SyntaxNode root, S
             var endToken = CreateInterpolatedStringEndToken()
                                 .WithTrailingTrivia(pieces.Last().GetTrailingTrivia());
 
-            var content = new List<SyntaxNode>();
-            foreach (var piece in pieces)
+            var content = new List<SyntaxNode>(pieces.Count);
+            var previousIndex = -2;
+            var insertionIndex = -1;
+            for (var i = 0; i < pieces.Count; i++)
             {
-                if (syntaxFacts.IsStringLiteralExpression(piece))
+                if (syntaxFacts.IsStringLiteralExpression(pieces[i]))
                 {
-                    var text = piece.GetFirstToken().Text;
+                    var text = pieces[i].GetFirstToken().Text;
                     var textWithoutQuotes = GetTextWithoutQuotes(text, isVerbatimStringLiteral);
-                    content.Add(generator.InterpolatedStringText(
-                        generator.InterpolatedStringTextToken(textWithoutQuotes)));
+                    if (i - 1 == previousIndex)
+                    {
+                        // Last part we added to the content list was also an interpolated-string-text-node.
+                        // We need to combine these as the API for creating an interpolated strings
+                        // does not expect to be given a list containing non-contiguous string nodes.
+                        // Essentially if we combine '"A" + 1 + "B" + "C"' into '$"A{1}BC"' it must be:
+                        //      {InterpolatedStringText}{Interpolation}{InterpolatedStringText}
+                        // not:
+                        //      {InterpolatedStringText}{Interpolation}{InterpolatedStringText}{InterpolatedStringText}
+                        var interpolatedStringTextNode = content[insertionIndex];
```

Comment:
I'm not thrilled with trying to do the mental gymnastics to prove that this never gets called with -1 or after the end of content.  Convince me (and future readers) that this is safe.